pkg/aws/dynamoblockindextable_test.go

```diff
@@ -0,0 +1,193 @@
+				"length":         &types.AttributeValueMemberN{Value: fmt.Sprint(length)},
+			},
+		})
+		require.NoError(t, err)
+
+		store := NewDynamoProviderBlockIndexTable(dynamoClient, tableName)
+
+		results, err := store.Query(ctx, digest)
+		require.NoError(t, err)
+		require.Equal(t, 1, len(results))
+		require.Equal(t, path, results[0].CarPath)
+		require.Equal(t, uint64(offset), results[0].Offset)
+		require.Equal(t, uint64(length), results[0].Length)
+	})
+
+	t.Run("query multiple existing items for same digest", func(t *testing.T) {
+		root := testutil.RandomCID()
+		digest := link.ToCID(root).Hash()
+
+		items := []struct {
+			path   string
+			offset int
+			length int
+		}{
+			{
+				path:   fmt.Sprintf("http://test.example.com/%s.blob", digestutil.Format(digest)),
+				offset: rand.IntN(1024 * 1024 * 128),
+				length: rand.IntN(1024*1024*2) + 1,
+			},
+			{
+				path:   fmt.Sprintf("http://test.example.com/%s.car", root.String()),
+				offset: rand.IntN(1024 * 1024 * 128),
+				length: rand.IntN(1024*1024*2) + 1,
+			},
+		}
+
+		for _, i := range items {
+			_, err := dynamoClient.PutItem(ctx, &dynamodb.PutItemInput{
+				TableName: aws.String(tableName),
+				Item: map[string]types.AttributeValue{
+					"blockmultihash": &types.AttributeValueMemberS{Value: digestutil.Format(digest)},
+					"carpath":        &types.AttributeValueMemberS{Value: i.path},
+					"offset":         &types.AttributeValueMemberN{Value: fmt.Sprint(i.offset)},
+					"length":         &types.AttributeValueMemberN{Value: fmt.Sprint(i.length)},
+				},
+			})
+			require.NoError(t, err)
+		}
+
+		store := NewDynamoProviderBlockIndexTable(dynamoClient, tableName)
+
+		results, err := store.Query(ctx, digest)
+		require.NoError(t, err)
+		require.Equal(t, len(items), len(results))
+
+		for _, i := range items {
+			require.True(t, slices.ContainsFunc(results, func(r legacy.BlockIndexRecord) bool {
+				return r.CarPath == i.path && r.Offset == uint64(i.offset) && r.Length == uint64(i.length)
+			}))
+		}
+	})
+
+	t.Run("query not found", func(t *testing.T) {
+		digest := testutil.RandomMultihash()
+		store := NewDynamoProviderBlockIndexTable(dynamoClient, tableName)
+		_, err := store.Query(ctx, digest)
+		require.Error(t, err)
+		require.True(t, errors.Is(err, istypes.ErrKeyNotFound))
+	})
+}
+
+func createDynamo(t *testing.T) url.URL {
+	ctx := context.Background()
+	req := testcontainers.ContainerRequest{
+		Image:        "amazon/dynamodb-local:latest",
+		ExposedPorts: []string{"8000/tcp"},
+		WaitingFor:   wait.ForExposedPort(),
+	}
+	container, err := testcontainers.GenericContainer(ctx, testcontainers.GenericContainerRequest{
+		ContainerRequest: req,
+		Started:          true,
+	})
+	testcontainers.CleanupContainer(t, container)
+	require.NoError(t, err)
+
+	endpoint, err := container.Endpoint(ctx, "")
+	require.NoError(t, err)
+
+	return *testutil.Must(url.Parse("http://" + endpoint))(t)
```

Comment:
Yeah I see...I'm just keen to communicating the fact that this isn't something that's going to be mutated.

Reply:
yes, I'm used to that too in other languages. The convention in Go is subtly different, I think because of how pointers are used in pointer receivers in functions to define the type's API. It feels a bit more idiomatic, but feel free to use whatever you prefer.